### PR TITLE
Additional GPIO functionality

### PR DESCRIPTION
**Pull request by fmckeogh**
* Input
* Alternate function
* Stateful
* Toggleable

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,8 +8,8 @@ edition = "2018"
 
 [dependencies]
 # esp32 = { version = "0.2.1" }
-esp32 = { path = "../../hal/esp32-pac"}
+esp32 = { path = "../esp32"}
```
Comment (jeandudey) [inline Cargo.toml]:
```suggestion
esp32 = { git = "https://github.com/esp-rs/esp32.git" }
```
and `cargo update` when necessary, until `esp32` gets released to crates.io.

Comment (MabezDev) [reply Cargo.toml]: I released 0.2.2 last night, https://crates.io/crates/esp32

File: src/gpio.rs [hunk 2]
```diff
@@ -166,40 +252,146 @@ macro_rules! impl_output {
...
+        Gpio27: (27, pin27, func27_out_sel_cfg, gpio27),
     ]
 }
 
 impl_output! {
-    enable1_w1ts, out1_w1ts, out1_w1tc, [
-        Gpio32: (0, pin32, func32_out_sel_cfg, gpio32, 0b00),
-        Gpio33: (1, pin33, func33_out_sel_cfg, gpio33, 0b00),
-        /* Deliberately omitting 34-39 as these can *only* be inputs */
+    enable_w1ts, out1_w1ts, out1_w1tc, [
```
Comment (MabezDev) [inline src/gpio.rs]:
```suggestion
    enable1_w1ts, out1_w1ts, out1_w1tc, [
```

File: src/gpio.rs [hunk 3]
```diff
@@ -166,40 +252,146 @@ macro_rules! impl_output {
...
+        Gpio26: (26, pin26, func26_out_sel_cfg, gpio26),
+        Gpio27: (27, pin27, func27_out_sel_cfg, gpio27),
     ]
 }
 
 impl_output! {
-    enable1_w1ts, out1_w1ts, out1_w1tc, [
-        Gpio32: (0, pin32, func32_out_sel_cfg, gpio32, 0b00),
-        Gpio33: (1, pin33, func33_out_sel_cfg, gpio33, 0b00),
-        /* Deliberately omitting 34-39 as these can *only* be inputs */
```
Comment (MabezDev) [inline src/gpio.rs:203]: Add the comment back please

File: src/gpio.rs [hunk 4]
```diff
@@ -166,40 +252,146 @@ macro_rules! impl_output {
...
+    enable_w1ts, out1_w1ts, out1_w1tc, [
+        Gpio32: (0, pin32, func32_out_sel_cfg, gpio32),
+        Gpio33: (1, pin33, func33_out_sel_cfg, gpio33),
+    ]
+}
+
+macro_rules! impl_input {
+    ($en:ident, $reg:ident, $reader:ident [
+        // index, gpio pin name, funcX name, iomux pin name, iomux mcu_sel bits
```
Comment (MabezDev) [inline src/gpio.rs]:
```suggestion
        // index, gpio pin name, funcX name, iomux pin name
```

File: src/gpio.rs [hunk 5]
```diff
@@ -42,6 +41,26 @@ pub struct Output<MODE> {
     _mode: PhantomData<MODE>,
 }
 
+/// Alternate function
+pub struct Alternate<MODE> {
+    _mode: PhantomData<MODE>,
+}
+
+/// Alternate Function 1
+pub struct Func1;
```
Comment (MabezDev) [inline src/gpio.rs]: Slightly superficial, but I don't really like `Func1`, could we have `AF1` etc instead?

File: src/gpio.rs [hunk 6]
```diff
@@ -166,40 +252,146 @@ macro_rules! impl_output {
...
+        Gpio22: (22, pin22, func22_in_sel_cfg, gpio22),
+        Gpio23: (23, pin23, func23_in_sel_cfg, gpio23),
+        Gpio25: (25, pin25, func25_in_sel_cfg, gpio25),
+        Gpio26: (26, pin26, func26_in_sel_cfg, gpio26),
+        Gpio27: (27, pin27, func27_in_sel_cfg, gpio27),
+    ]
+}
+
+impl_input! {
+    enable_w1ts, in1, in1_data [
```
Comment (MabezDev) [inline src/gpio.rs]:
```suggestion
    enable1_w1ts, in1, in1_data [
```

**Review by MabezDev — CHANGES_REQUESTED** [review]
Looking good! Will test on a board asap.

File: src/gpio.rs (hunk 3 above)
Comment (fmckeogh) [reply src/gpio.rs:203]: My bad, I didn't mean to remove that

**Comment by fmckeogh** [comment]
All fixed! :)

**Comment by MabezDev** [comment]
Thanks! I tested some input pins last night on a real board and it worked great! Once I test some pins in the second bank (32+) I'll merge this. Alternate function looks good too, but I guess we don't really have a way of testing these until we write some drivers for other peripherals.

**Comment by fmckeogh** [comment]
Yeah, I didn't have a great way of testing pull up/down vs floating other than setting the LED to the state read from the pin and looking for an immediate change in LED state vs a slow fade, but it does appear to be correct.

I've started on I2C, and wasn't someone else working on serial? So shouldn't be long! :)

**Comment by arjanmels** [comment]
@chocol4te There are still a couple of features which I think will be needed for the library to become usable for all situations:
- implement alternate mode for inputs (with configurable pull-up/down?)
- implement setting of routing via GPIO to peripherals (via GPIO_FUNCm_IN_SEL_CFG_REG and GPIO_FUNCm_OUT_SEL_CFG_REG registers)

Is this something you are working on/will be working on? (can be next pull request. I think the current one is already very nice!)

**Comment by fmckeogh** [comment]
Sorry, I've been pretty busy with university work, I meant to add both of those features to a pull request implementing I2C but haven't had time recently :(

**Comment by arjanmels** [comment]
@chocol4te No worries, it is just a hobby, other things have prio. For the moment I work around it by just using UART0 with the default pin assignment and will await your updates in due time.

**Comment by MabezDev** [comment]
Sorry for taking so long to get to this @chocol4te!

This looks great, I've tested most pins from each bank in various number of scenarios and all seems to be working great.

One thing I have noticed is that pullups/pulldowns are not available for pins 34-39, so calling into_x_pullup doesn't have the intended effect. I will still merge this but hopefully we can find away to make it such that those functions are not on those pins in the near future!

**Pull request merged**